src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.sql

```diff
@@ -1629,6 +1689,58 @@ AS
     COMMIT TRANSACTION
 GO
 
+/***************************************************************************************/
+-- STORED PROCEDURE
+--     AddExtendedQueryTagError
+--
+-- DESCRIPTION
+--    Adds an Extended Query Tag Error or Updates it if exists.
+--
+-- PARAMETERS
+--     @tagKey
+--         * The related extended query tag's key
+--     @errorCode
+--         * The error code
+--     @watermark
+--         * The watermark
+--
+-- RETURN VALUE
+--     The tag key of the error added.
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.AddExtendedQueryTagError (
+    @tagKey INT,
+    @errorCode INT,
+    @watermark BIGINT
+)
+AS
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+    BEGIN TRANSACTION
+
+    DECLARE @currentDate DATETIME2(7) = SYSUTCDATETIME()
+
+        --Check if instance exists
+        IF NOT EXISTS (SELECT * FROM dbo.Instance WITH (UPDLOCK) WHERE Watermark = @watermark AND Status = 1)
+            THROW 50404, 'Instance does not exist or has not been created', 1;
+
+        --Check if tag exists
+        IF NOT EXISTS (SELECT * FROM dbo.ExtendedQueryTag WITH (HOLDLOCK) WHERE TagKey = @tagKey)
```

Comment:
Tags too have statuses and should only be considered if it's in the process of being added. However, we've been using the presence of an operation to indicate whether the tag is "pending" vs "adding."
@pengchen0692 , while our state machine doesn't allow for adds to happen in this order, I am wondering whether we should create a separate tag status for "pending" to make querying easier.